nuxeo-core/nuxeo-core-bulk/src/main/java/org/nuxeo/ecm/core/bulk/BulkServiceImpl.java

```diff
@@ -114,4 +118,29 @@ public KeyValueStore getKvStore() {
         return Framework.getService(KeyValueService.class).getKeyValueStore(BULK_KV_STORE_NAME);
     }
 
+    @Override
+    public boolean await(Duration duration) throws InterruptedException {
+        StreamService service = Framework.getService(StreamService.class);
+        LogManager logManager = service.getLogManager(BULK_LOG_MANAGER_NAME);
+        BulkAdminService admin = Framework.getService(BulkAdminService.class);
+        Collection<String> actions = admin.getActions();
+        Collection<String> streams = new ArrayList<>(actions.size() + 3);
+        streams.add(DOCUMENTSET_ACTION_NAME);
+        streams.addAll(actions);
+        streams.add(COUNTER_ACTION_NAME);
+        streams.add(KVWRITER_ACTION_NAME);
+        long deadline = System.currentTimeMillis() + duration.toMillis();
+        for (String stream : streams) {
+            // when there is no lag between producer and consumer we are done
+            while (logManager.getLag(stream, stream).lag() > 0) {
+                if (System.currentTimeMillis() > deadline) {
+                    return false;
+                }
+                Thread.sleep(50);
+            }
+            // we wait for records to be actually passed to next stream
+            Thread.sleep(100);
```

Comment:
Could you do this sleep only once for all streams, not once per stream?

Reply:
Actually no, the point is when a computation produce a record to another stream, it may have a little delay before the following computation treats it, we have to sleep here to ensure the check does not pass between the 2 computations.